router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+			if client, ok := p.clients[topic]; ok {
+				client.Close()
+				delete(p.clients, topic)
+			}
+		} else {
+			p.subscriptions[topic] = tmp
+		}
+		p.mu.Unlock()
+	}
+}
+
+// createClient creates a new client for the given topic if it does not exist yet.
+// It returns true if the client was created, false otherwise.
+func (p *kafkaPubSub) createClient(topic string) (*kgo.Client, bool, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if _, ok := p.clients[topic]; ok {
+		return nil, false, nil
+	}
+
+	// Create a new client for the topic
+	client, err := kgo.NewClient(append(p.opts,
+		kgo.ConsumeTopics(topic),
+		// We want to consume the events produced after the first subscription was created
+		// Messages are shared among all subscriptions, therefore old events are not redelivered
+		// This replicates a stateless publish-subscribe model
+		kgo.ConsumeResetOffset(kgo.NewOffset().AfterMilli(time.Now().UnixMilli())),
+		// For observability, we set the client ID to "router"
+		kgo.ClientID(fmt.Sprintf("cosmo.router.consumer.%s", topic)),
+	)...)
+	if err != nil {
+		return nil, false, fmt.Errorf(`failed to create client for Kafka for topic "%s": %w`, topic, err)
+	}
+
+	p.clients[topic] = client
+
+	return client, true, nil
+}
+
+// Publish publishes the given event to the Kafka topic in a non-blocking way.
+// Publish errors are logged and returned as a pubsub error.
+// The event is written with a dedicated write client.
+func (p *kafkaPubSub) Publish(ctx context.Context, event pubsub_datasource.KafkaPublishEventConfiguration) error {
+
+	p.logger.Debug("publish",
+		zap.String("topic", event.Topic),
+		zap.String("providerID", event.ProviderID),
+		zap.ByteString("data", event.Data),
+	)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var pErr error
+
+	p.writeClient.Produce(ctx, &kgo.Record{
+		Topic: event.Topic,
+		Value: event.Data,
+	}, func(record *kgo.Record, err error) {
+		defer wg.Done()
+		if err != nil {
+			pErr = err
+		}
+	})
+
+	wg.Wait()
+
+	if pErr != nil {
+		p.logger.Error("publish error",
+			zap.String("topic", event.Topic),
+			zap.String("providerID", event.ProviderID),
+			zap.Error(pErr),
+		)
+		return pubsub.NewError(fmt.Sprintf("error publishing to Kafka topic %s", event.Topic), pErr)
+	}
+
+	return nil
+}
+
+func (p *kafkaPubSub) flush(ctx context.Context) error {
+	err := p.writeClient.Flush(ctx)
+	if err != nil {
+		p.logger.Error("error flushing write client", zap.Error(err))
+	}
+
+	for _, client := range p.clients {
+		if err := client.Flush(ctx); err != nil {
+			p.logger.Error("error flushing client", zap.Error(err))
+		}
+	}
+
+	return nil
+}
+
+func (p *kafkaPubSub) Shutdown(ctx context.Context) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var err error
+
+	fErr := p.flush(ctx)
+	if fErr != nil {
+		p.logger.Error("error flushing clients", zap.Error(err))
+		err = errors.Join(err, fErr)
+	}
+
+	p.writeClient.Close()
+
+	for _, client := range p.clients {
+		client.Close()
```

Comment:
You already close the clients when p.ctx.Done() is called in the poller. What happens when client.Close gets called twice?
If you'd delete the client from clients after close, with the mutex, would this prevent the issue?

Reply:
Nothing will happen. It can be called multiple times.